### PR TITLE
290 297 bulk generate serialize

### PR DESCRIPTION
**Pull request by CamBuscher**
This modal combines the work that Jonathan helped us do with the serializing of data using Jbuilder, and the Bulk Generate modal functionality. 

Serializing data allowed for the taking out of some unnecessarily verbose methods, which were full of mapping through deep object navigation, and made the manipulating of state much simpler.

Plus, the simpler organization of state made it possible to adjust the response of the Update action on the time_slots_controller so that state wasn't updated until the API responded successfully. Now the Update call responds with correct data. 

Finally, I added some tweaks to support other browsers CSS issues with the layout.

### CODE REVIEW

File: app/javascript/components/Schedule.js [hunk 1]
```diff
@@ -58,63 +60,77 @@ class Schedule extends React.Component {
...
-  };
+    postBulkTimeSlots(bulkPath, day, rooms, duration, formattedTimes, csrf)
+      .then(response => response.json())
+      .then(data => this.setState({ slots: data.slots, previewSlots: [] }))
+      .catch(err => console.log('Error: ', err))
+  }
 
   componentDidMount() {
-    let hours = this.determineHours(this.props.schedule.slots);
+    console.log(this.props)
```
Comment (jonsgreen) [inline app/javascript/components/Schedule.js]: Maybe you don't need this anymore.

File: app/views/staff/grids/schedule.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+json.rooms current_event.rooms, :id, :name
+json.days (1..current_event.days).to_a
+json.slots current_event.time_slots, :conference_day, :id, :update_path, :start_time, :end_time, :program_session_id, :track_id, :room_id
+json.sessions current_event.program_sessions
+json.unscheduledSessions current_event.program_sessions.unscheduled
+json.counts EventStats.new(current_event).schedule_counts
+json.tracks current_event.tracks.sort_by_name
+json.csrf session[:_csrf_token]
+json.bulkPath Rails.application.routes.url_helpers.event_staff_schedule_grid_bulk_time_slot_path(current_event)
```
Comment (jonsgreen) [inline app/views/staff/grids/schedule.json.jbuilder]: I am not sure but you may be already have access to all url_helpers here so you can try just making this line: 
```
json.bulkPath event_staff_schedule_grid_bulk_time_slot_path(current_event)
```

**Review by jonsgreen — APPROVED** [review]
A am not quite proficient enough with React and Javascript to authorize but my sense is that this is a fine piece of work and a certainly a very dedicated effort to be proud of.

File: app/controllers/staff/grids_controller.rb [hunk 3]
```diff
@@ -2,10 +2,10 @@ class Staff::GridsController < Staff::ApplicationController
   include ScheduleSupport
 
   def show
-    @schedule = Schedule.new(current_event)
-    @counts = EventStats.new(current_event).schedule_counts
-    @sessions = current_event.program_sessions
-    @unscheduled_sessions = @sessions.unscheduled
-    @tracks = current_event.tracks.sort_by_name
+    # @schedule = Schedule.new(current_event)
```
Comment (mghaught) [inline app/controllers/staff/grids_controller.rb]: If we're not planning to use these, let's delete the lines.  We can always use git to gain access if we need them for some reason.   Better to remove than leave them commented out for future devs to wonder.

File: app/views/staff/grids/schedule.json.jbuilder (hunk 2 above)
Comment (CamBuscher) [reply app/views/staff/grids/schedule.json.jbuilder]: Your hunch was correct! Refactored for the next push.

File: app/controllers/staff/grids_controller.rb (hunk 3 above)
Comment (CamBuscher) [reply app/controllers/staff/grids_controller.rb]: Thank you for this, addressed in this push I am deploying.

**Comment by CamBuscher** [comment]
Redeployed with changes after discussions with Jeneve & Marty. Major ones listed below, plus a ton of minor code style choices (eliminating semicolons, name issues, etc).

- Unscheduled area scrolls with the grid both horizontally and vertically.
- Added a hover effect on slots when dragging a session, this will help with attaching a session to a small slot. 
- Preview view for slots more resembles original, and confirmation box is near the generate grid button.
- Modal for generate grid requires information before preview, denotes what fields are required.
- Modal allows for selecting time based on formats.
- Grid resizes if a preview is outside of previously defined hours, and extends hours on creation of time slot.

**Pull request merged**